staging/src/k8s.io/sample-controller/integration_test.go

```diff
@@ -0,0 +1,215 @@
+	//		DeploymentName: "example-foo",
+	//		Replicas:       &replicas,
+	//	},
+	// }
+	// _, err = exampleClient.SamplecontrollerV1alpha1().Foos(ns.Name).Create(exampleFoo)
+	// if err != nil {
+	//	t.Fatalf("error creating example foo: %v", err)
+	// }
+
+	// Create an instance of the CRD using dynamic client.
+	// Following example from https://github.com/kubernetes/apiextensions-apiserver/blob/release-1.8/test/integration/registration_test.go#L42
+	exampleFoo := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "samplecontroller.k8s.io/v1alpha1",
+			"kind":       "Foo",
+			"metadata": map[string]interface{}{
+				"namespace": ns.Name,
+				"name":      "example-foo",
+			},
+			"spec": map[string]interface{}{
+				"deploymentName": "example-foo",
+				"replicas":       123,
+			},
+		},
+	}
+	resourceClient := dynamicClient.Resource(
+		&metav1.APIResource{
+			Name:       crd.Spec.Names.Plural,
+			Namespaced: crd.Spec.Scope != apiextensionsv1beta1.ClusterScoped,
+		},
+		ns.Name,
+	)
+	_, err = resourceClient.Create(exampleFoo)
+	if err != nil {
+		t.Fatalf("error creating example foo: %v", err)
+	}
+
+	// The example client can find the created Foo resource
+	found, err := exampleClient.SamplecontrollerV1alpha1().Foos(ns.Name).Get(
+		"example-foo",
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		t.Fatalf("error getting example-foo: %v", err)
+	}
+	t.Logf("Found example-foo: %v", found)
+
+	// Create and start the sample-controller and its informers
+	kubeInformerFactory := kubeinformers.NewSharedInformerFactory(kubeClient, 0)
+	exampleInformerFactory := informers.NewSharedInformerFactory(exampleClient, 0)
+
+	controller := NewController(
+		kubeClient,
+		exampleClient,
+		kubeInformerFactory,
+		exampleInformerFactory,
+	)
+
+	stopCh := make(chan struct{})
+	controllerFinished := make(chan struct{})
+	defer func() {
+		close(stopCh)
+		<-controllerFinished
+	}()
+
+	kubeInformerFactory.Start(stopCh)
+	exampleInformerFactory.Start(stopCh)
+
+	go func() {
+		defer close(controllerFinished)
+		err := controller.Run(2, stopCh)
+		if err != nil {
+			t.Fatalf("error running controller: %v", err)
+		}
+	}()
+
+	// Wait 60s for the controller to create a deployment
+	err = wait.PollImmediate(
+		5*time.Second,
+		60*time.Second,
+		func() (bool, error) {
+			deployment, err := kubeClient.AppsV1beta2().Deployments(ns.Name).Get(
+				"example-foo",
+				metav1.GetOptions{},
+			)
+			if err != nil {
+				if errors.IsNotFound(err) {
+					return false, nil
+				}
+				return false, err
+			}
+			if *deployment.Spec.Replicas != 123 {
+				return false, fmt.Errorf(
+					"deployment had unexpected replicas value: %v",
+					*deployment.Spec.Replicas,
+				)
+			}
+			return true, nil
+		},
+	)
+	if err != nil {
+		t.Fatalf("controller failed to create expected deployment: %v", err)
+	}
+
+	// TODO:
+	// * Test for recorded events
+	// * Test with existing deployments
+	// * Test Foo status updates.
```

Comment:
When it comes to writing these tests, it'd be ideal to have them in their own functions (to make the `go test` output more useful, as well as provided an example of how others can begin writing their own similar tests).

I think some parts of this function will need to be either generalised or moved into some common 'beforeTest' function to do so.